### PR TITLE
Added location option to test target

### LINKED ISSUE #1150: Location Simulation For Test Targets
Hi All!

In a previous issue #722 & #756, Location Simulation was requested and added as a feature in XcodeGen. Is it possible to expand upon this and include location within test targets as well?

Currently, the way to simulate location is by adding the lines below, however, this is limited to normal runs in the simulator and does not work inside a UI test run:
```
run:
  simulateLocation:
     allow: true
     defaultLocation: Honolulu, HI, USA
```

I'm wondering if it's possible to create something like below. I added an image below to show how this would look in the build scheme:
```
test:
      gatherCoverageData: true
      targets:
        - App_iOS_UITests
        - name: App_iOS_Tests
          parallelizable: true
          randomExecutionOrder: true
          location: Honolulu, HI, USA  ### FEATURE REQUEST
```

Screen Shot 2021-11-22 at 1 02 38 PM

Screen Shot 2021-11-23 at 10 41 37 AM

### PR DESCRIPTION
**Pull request by KrisRJack**
Resolves #1150

### CODE REVIEW

File: Sources/XcodeGenKit/SchemeGenerator.swift [hunk 1]
```diff
@@ -187,12 +187,24 @@ public class SchemeGenerator {
...
+                locationScenarioReference = XCScheme.LocationScenarioReference(identifier: identifier, referenceType: referenceType.rawValue)
+            }
+            
+            return XCScheme.TestableReference(
                 skipped: testTarget.skipped,
                 parallelizable: testTarget.parallelizable,
                 randomExecutionOrdering: testTarget.randomExecutionOrder,
-                buildableReference: testBuilEntries.buildableReference,
+                buildableReference: testBuildEntries.buildableReference,
+                locationScenarioReference: locationScenarioReference,
```
Comment (KrisRJack) [inline Sources/XcodeGenKit/SchemeGenerator.swift:207]: Created PR for this here in XcodeProj: https://github.com/tuist/XcodeProj/pull/654

**Comment by KrisRJack** [comment]
Hey All! Let me know if there is anymore that I can do here on my end to help move this PR forward. Thank you! (cc @yonaskolb)

**Comment by farshadtx** [comment]
@yonaskolb can we get a little love to this one?!

**Comment by LuAndreCast** [comment]
@yonaskolb can we get this fixed please?

File: Sources/ProjectSpec/Scheme.swift [hunk 2]
```diff
@@ -184,13 +184,43 @@ public struct Scheme: Equatable {
         public var deleteScreenshotsWhenEachTestSucceeds: Bool
 
         public struct TestTarget: Equatable, ExpressibleByStringLiteral {
+            
+            public struct SimulateLocation: Equatable {
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Scheme.swift:188]: Let's rename this to `Location`

File: Sources/ProjectSpec/Scheme.swift [hunk 3]
```diff
@@ -184,13 +184,43 @@ public struct Scheme: Equatable {
...
+                
+                public enum ReferenceType: String {
+                    case predefined = "1"
+                    case gpx = "0"
+                }
+                
+                public var location: String?
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Scheme.swift:195]: Any reason this needs to be optional?

File: Sources/ProjectSpec/Scheme.swift [hunk 4]
```diff
@@ -184,13 +184,43 @@ public struct Scheme: Equatable {
...
+                    }
+                    
+                    if location.contains(".gpx") {
+                        return .gpx
+                    }
+                    
+                    return .predefined
+                }
+                
+                public init(location: String) {
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Scheme.swift:210]:
```suggestion
                public init(_ location: String) {
```

File: Sources/ProjectSpec/Scheme.swift [hunk 5]
```diff
@@ -559,6 +593,9 @@ extension Scheme.Test.TestTarget: JSONEncodable {
         if parallelizable != Scheme.Test.TestTarget.parallelizableDefault {
             dict["parallelizable"] = parallelizable
         }
+        if let location = location {
+            dict["location"] = location
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Scheme.swift:597]: This just puts the struct into the dictionary. Write the string
```suggestion
            dict["location"] = location.location
```

File: Sources/ProjectSpec/Scheme.swift [hunk 6]
```diff
@@ -536,6 +569,7 @@ extension Scheme.Test.TestTarget: JSONObjectConvertible {
         targetReference = try TargetReference(jsonDictionary.json(atKeyPath: "name"))
         randomExecutionOrder = jsonDictionary.json(atKeyPath: "randomExecutionOrder") ?? Scheme.Test.TestTarget.randomExecutionOrderDefault
         parallelizable = jsonDictionary.json(atKeyPath: "parallelizable") ?? Scheme.Test.TestTarget.parallelizableDefault
+        location = jsonDictionary.json(atKeyPath: "location") ?? nil
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Scheme.swift:572]: Decode the location as a string. The spec says it's a string. At the moment you would need:
```
location:
  location: abc.gpx
```

File: Docs/ProjectSpec.md [hunk 7]
```diff
@@ -834,6 +834,7 @@ A multiline script can be written using the various YAML multiline methods, for
 - [x] **name**: **String** - The name of the target
 - [ ] **parallelizable**: **Bool** - Whether to run tests in parallel. Defaults to false
 - [ ] **randomExecutionOrder**: **Bool** - Whether to run tests in a random order. Defaults to false
+- [ ] **location**: **String** - GPX file or predefined value for simulating location. See [Simulate Location](#simulate-location).
```
Comment (yonaskolb) [inline Docs/ProjectSpec.md:837]:
```suggestion
- [ ] **location**: **String** - GPX file or predefined value for simulating location. See [Simulate Location](#simulate-location) for location examples.
```

**Review by yonaskolb — COMMENTED** [review]
Thanks for this @KrisRJack. Left some comments, mainly around simplifying the location spec to a string instead of an object.
Can you also:
- add a decoding test to SpecLoadingTests
- add a generator test to SchemeGeneratorTests
- add an example to the Fixtures/TestProject/project.yml

**Comment by KrisRJack** [comment]
Hey Yonas! Thank you so much for the feedback! I'll have these changes in shortly.

**Comment by KrisRJack** [comment]
Hey @yonaskolb! Thanks again for the feedback and the review. Below, I've outlined the changes made to this PR:

- I removed the `SimulateLocation` struct entirely and just passed in a string of the location 
- Removed `SimulateLocation` from SchemeGenerator.swift and gathered location reference type and identifier from given string
- Expanded upon existing tests in SpecLoadingTests and SchemeGeneratorTests to include location in test targets. The commits test both for custom `.gpx` files and default locations such as "New York, NY, USA" 
- Added an example to the Fixtures/TestProject/project.yml

Thanks again and let me know if there's anything more that I can add to this.

**Review by yonaskolb — APPROVED** [review]
Fantastic! Thanks again!

**Comment by KrisRJack** [comment]
Hey @yonaskolb! Thanks again for your review and merging! Really awesome to see this in the project. 

Any chance you can update the release to 2.26.0 as well so our team can apply this feature to our repo? Thanks again!

**Comment by yonaskolb** [comment]
@KrisRJack released 👍

**Pull request merged**